### ISSUE TITLE
Inputs broken

### ISSUE THREAD

**Issue by mahoneyt944** [libretro/mame2003-plus-libretro#1059]
@markwkidd  After lastest commits.

Mame menu won't stay open. Couldn't test further.

**Comment by mahoneyt944** [comment]
Looks like the b button is auto activated on every quick key press

**Comment by markwkidd** [comment]
Ugh. I will check in a moment. Very sorry

**Comment by markwkidd** [comment]
@mahoneyt944 could you clear your CFG files and try again if you haven't? I am not seeing this problem with the MAME menu or the spammed B button locally.

**Comment by mahoneyt944** [comment]
Just retested. Issue still exists after deleting the default.cfg and my game.cfg. when I press any key quickly it activates that key followed by a b button press.

**Comment by markwkidd** [comment]
I'm running commit 4554e155 with Super Street Fighter 2 Turbo and having trouble getting this to happen. I feel like I'm missing something obvious.

I can start the game, and throw some punches and kicks, and things seem to look right. I can press TAB to get into the MAME menu and remap things. Thank you for your time on this.

**Comment by mahoneyt944** [comment]
For me it has to be a quick press. A slow press won't do it. I'm testing with baluba. I can press any button, even ones not mapped to anything, and I can can get him to jump. What seems to happen is the pressed button activates first then b automatically. So if I tap left quickly, baluba will turn left and jump

**Comment by grant2258** [comment]
I just compiled this and tried it on linux cant reproduce can you post your core options file and ill try run it with the same as yours.

**Comment by mahoneyt944** [comment]
I just deleted the opt file so I'm using completely default right now. On Android with on screen retropad 1.9.3 retroarch

**Comment by mahoneyt944** [comment]
Seems if I try to map in the mame menu "gun1 trigger" activates alot. Which shouldn't be at all.

**Comment by grant2258** [comment]
might be something to do wit this https://github.com/libretro/RetroArch/commit/7174563f7d09b55de0dcd5c2f8fcc9c71674fea4

**Comment by mahoneyt944** [comment]
I'll try downgrading to 1.9.2

**Comment by mahoneyt944** [comment]
Well after f-ing up my setup, uninstalling, and then  installing 1.9.3 again. I tested mame2003 and the issue is there too. So seems to be retroarch.

**Comment by mahoneyt944** [comment]
Ok I fixed it in 2003 by turning off the mouse option. Still in issue in 2003+ though.

**Comment by mahoneyt944** [comment]
I believe the issue is that the new gun code is activated when then touch screen is pressed. Even when the core option is off

**Comment by markwkidd** [comment]
I'm not familiar with touch screens + RetroArch, but is part of the answer to create a core option to disable lightgun polling, in the same way mouse can be disabled? I have been thinking a 'lightgun enable/disable' core option might be good to add for debugging and testing anyway.

What you are describing doesn't seem like it would be the intended RA behavior with a touch overlay though.

**Comment by mahoneyt944** [comment]
The touch screen is just a mouse essentially. I would use the mouse core option we have already. When it's set to "off" the gun shouldn't work either.

**Comment by mahoneyt944** [comment]
I think the options get too confusing to end users. Why not just add "light gun" under mouse device core option? And poll the gun if the option is selected?  That way the gun is available for all controller types and the user can decide to use a mouse, pointer, lightgun or none at all?

**Comment by mahoneyt944** [comment]
So basically:
```
        case OPT_MOUSE_DEVICE:
          if(strcmp(var.value, "pointer") == 0)
            options.mouse_device = RETRO_DEVICE_POINTER;
          else if(strcmp(var.value, "mouse") == 0)
            options.mouse_device = RETRO_DEVICE_MOUSE;
          else if(strcmp(var.value, "lightgun") == 0)
            options.mouse_device = RETRO_DEVICE_LIGHTGUN;
          else
            options.mouse_device = RETRO_DEVICE_NONE;
          break;
```

Then we integrate lightgun into our mouse options here so it's polled when selected
```
    /* do not poll mouse abstraction when disabled by the core option or if user explicitly selects Lightgun */
    if(options.mouse_device != RETRO_DEVICE_NONE && get_device_parent(device_type) != RETRO_DEVICE_LIGHTGUN)
    {
      if(options.mouse_device == RETRO_DEVICE_MOUSE)
      {
        retroJsState[port][OSD_MOUSE_BUTTON_1]  = input_cb(port, RETRO_DEVICE_MOUSE, 0, RETRO_DEVICE_ID_MOUSE_LEFT);
        retroJsState[port][OSD_MOUSE_BUTTON_2]  = input_cb(port, RETRO_DEVICE_MOUSE, 0, RETRO_DEVICE_ID_MOUSE_RIGHT);
        retroJsState[port][OSD_MOUSE_BUTTON_3]  = input_cb(port, RETRO_DEVICE_MOUSE, 0, RETRO_DEVICE_ID_MOUSE_MIDDLE);
        retroJsState[port][OSD_MOUSE_BUTTON_4]  = input_cb(port, RETRO_DEVICE_MOUSE, 0, RETRO_DEVICE_ID_MOUSE_BUTTON_4);
        retroJsState[port][OSD_MOUSE_BUTTON_5]  = input_cb(port, RETRO_DEVICE_MOUSE, 0, RETRO_DEVICE_ID_MOUSE_BUTTON_5);
        mouse_x[port] = input_cb(port, RETRO_DEVICE_MOUSE, 0, RETRO_DEVICE_ID_MOUSE_X);
        mouse_y[port] = input_cb(port, RETRO_DEVICE_MOUSE, 0, RETRO_DEVICE_ID_MOUSE_Y);
      }
      else if (options.mouse_device == RETRO_DEVICE_POINTER)
      {
        bool pointer_pressed = input_cb(port, RETRO_DEVICE_POINTER, 0, RETRO_DEVICE_ID_POINTER_PRESSED);
        retroJsState[port][OSD_MOUSE_BUTTON_1]  = pointer_pressed;
        mouse_x[port] = pointer_pressed ? get_pointer_delta(input_cb(port, RETRO_DEVICE_POINTER, 0, RETRO_DEVICE_ID_POINTER_X), &prev_pointer_x) : 0;
        mouse_y[port] = pointer_pressed ? get_pointer_delta(input_cb(port, RETRO_DEVICE_POINTER, 0, RETRO_DEVICE_ID_POINTER_Y), &prev_pointer_y) : 0;
      }
    }

    /* poll lightgun */
    lightgun_x[port] = normalize_lightgun(input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_SCREEN_X));
    lightgun_y[port] = normalize_lightgun(input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_SCREEN_Y));
    retroJsState[port][OSD_LIGHTGUN_IS_TRIGGER]  = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_TRIGGER); /*Status Check*/
    retroJsState[port][OSD_LIGHTGUN_RELOAD]      = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_RELOAD);  /*Forced off-screen shot*/
    retroJsState[port][OSD_LIGHTGUN_AUX_A]       = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_AUX_A);
    retroJsState[port][OSD_LIGHTGUN_AUX_B]       = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_AUX_B);
    retroJsState[port][OSD_LIGHTGUN_START]       = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_START);
    retroJsState[port][OSD_LIGHTGUN_SELECT]      = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_SELECT);
    retroJsState[port][OSD_LIGHTGUN_AUX_C]       = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_AUX_C);
    retroJsState[port][OSD_LIGHTGUN_DPAD_UP]     = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_DPAD_UP);
    retroJsState[port][OSD_LIGHTGUN_DPAD_DOWN]   = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_DPAD_DOWN);
    retroJsState[port][OSD_LIGHTGUN_DPAD_LEFT]   = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_DPAD_LEFT);
    retroJsState[port][OSD_LIGHTGUN_DPAD_RIGHT]  = input_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_DPAD_RIGHT);

    /* simulated lightgun reload hack */
    if(device_parent == RETRO_DEVICE_LIGHTGUN)
    {
      if(retroJsState[port][OSD_LIGHTGUN_RELOAD])
      {
        retroJsState[port][OSD_LIGHTGUN_IS_TRIGGER] = true;
        lightgun_x[port] = -128;
        lightgun_y[port] = -128;
      }
    }
  }
```

**Comment by grant2258** [comment]
My only thoughts is why is android even setting  retroJsState[port][OSD_LIGHTGUN_RELOAD]

**Comment by mahoneyt944** [comment]
I know the mouse acts the same way which is why you have to turn it off with touch screen

**Comment by grant2258** [comment]
yea turn the mouse off in plus and see what code the tab menu give you to see if its a gun trigger or something elsewhere

**Comment by mahoneyt944** [comment]
Yeah its the trigger. It treats the screen touch the same as a mouse click or lightgun trigger. Here's my proposal. https://github.com/libretro/mame2003-plus-libretro/pull/1060/files

basically have the user decide which they want to use. We could then eliminate the lightgun controller type and just leave our alternate types.

**Comment by grant2258** [comment]
https://github.com/libretro/mame2003-plus-libretro/blob/4554e155bb67661145f31a8ac68f68bfb872a3fa/src/mame2003/mame2003.c#L1322

simple test do this and only apply to light gun games

if(device_parent == RETRO_DEVICE_LIGHTGUN && options.content_flags[CONTENT_LIGHTGUN])

**Comment by mahoneyt944** [comment]
Well I think ideally we should remove lightgun as a device parent all together. Let the gamepad, fight stick, etc be the types to pick from then the user can decide what xy device they want to poll in the core option? Seems simple enough. The issue with enabling it by content would be that with a touch screen in use as a gamepad, I can never use is as a lightgun too without messing up the on screen gamepad.

**Comment by grant2258** [comment]
but it wont work you just disabled the polling if selected as a device

**Comment by mahoneyt944** [comment]
> but it wont work you just disabled the polling if selected as a device

In the current setup yeah. You'd have to select lightgun as a device and select lightgun as the xy core option.

That's why I said we should remove it as a device type and just use it when selected as a core option. It would need more work done.

**Comment by grant2258** [comment]
This is why input need more thought in lightgun games it should do this on a desktop pc for a mouse and real lightgun. On andriod it should only do it on lightgun games are running The solution is simple do and andriod #ifdef then check the lightgun contents if false set all mouse and gun buttons in retrojs to 0

**Comment by mahoneyt944** [comment]
This issue extends past android though to any device with a touch screen. If the user wants to use the screen as an on screen retropad then it wouldn't make sense to poll a mouse, pointer, or lightgun since the screens already in use and would give false clicks / presses / shots etc.

I think if we just limited the device port types to gamepad, fightstick, 8button, 6button only...then we allow the user to decide in the core options which X-Y device they want to poll then everything would be supported across the board. The user could turn it completely off in the core options when using a touch screen or they could turn it on with any of the device port types.

**Comment by grant2258** [comment]
If you choose not to poll lightguns or mouse how would you get the pointer aimed on android for lightgun games thats the only issue i can see and there is no generic way for get this info from RA that know of you tell it what your using.

**Comment by mahoneyt944** [comment]
you use the on screen retropad. it acts the same way as a normal retropad. I play revx all the time using the onscreen analog stick.

**Comment by grant2258** [comment]
thing you need to keep in mind using this does not cancel the touchscreen/gun functionality so that why i told you to use a ifdefine to cancel the mouse/gun buttons if its android and not a lightgun game. There is no way to detect your using the screen pad so you need to do the logic for the platfrom you using

**Comment by markwkidd** [comment]
I had to sleep for a while, but I'm waking back up :)

I've never used RA with a touchscreen, but I still think something sounds wrong with its touchscreen overlay. Nonetheless there should be some way to cut off the lightgun from being polled. 

I've come around to the idea that it should be incorporated into the existing mouse interface core option.

@mahoneyt944 I agree with the plan of removing the gun as a controller type. I was only waiting until I had these new input codes set up to remove it, even before this issue. 

Up until a few weeks ago I would have said different because I intended to make more use of the controller type libretro functionality, for example I was even trying out setting up Trackball and Spinner controllers similar to what FB Neo does, but I found RetroArch's controller type selection implementation is still immature -- there were too many issues on the frontend to make a good user experience.

@grant2258 as you're pointing out, there is still a lot more that could be done to make this smarter. I would much rather only send default gun bindings to MAME for games with a gun.... maybe that is the next project
